File: examples/cpp/_3_Struct/subscriber.cpp
#include <iostream>
#include <thread>
#include "SharedPubSub.hpp"
#include "Examples.h"
using namespace std;

int main(){

    shps::Subscriber<ExampleStruct> subscriber("Struct", "StructSubscriber1",true);
    
    while(1){
        /*--------- Example 1 : Wait for notification ---------*/
        optional<ExampleStruct> exampleStruct1 = subscriber.readWait();
        if(exampleStruct1.has_value()){
            cout << "SUBSCRIBER : " << dec << 
            exampleStruct1.value().value1 << ", " <<
            exampleStruct1.value().value2 << ", " <<
            exampleStruct1.value().value3 << " " <<
            endl;
        }
        else{
            /*--------- Example 2 : Nothing was in queue, it was a normal notify, read the current value (not thread safe here)---------*/
            ExampleStruct exampleStruct2 = subscriber.readValue();
            cout << "SUBSCRIBER : " << dec << 
            exampleStruct2.value1 << ", " <<
            exampleStruct2.value2 << ", " <<
            exampleStruct2.value3 << " CURRENT VALUE" <<
            endl;
        }
    }
    return 0;
}

File: examples/nodejs/_4_ObjectAtomic/subscriber.js
const SharedPubSub = require('../build/Release/SharedPubSub');

const subscriber = new SharedPubSub.Subscriber_Object("AtomicObject", "AtomicObjectSubscriberJS", true);
var atomicObject;

async function loop() {
    while(1){
        /*--------- Example 1 : Wait for notification ---------*/
        atomicObject = subscriber.readWait();
        // Verify if the queue had a value on notification.
        if(atomicObject !== null && atomicObject !== undefined){
            console.log(`SUBSCRIBER : Value1: ${atomicObject.value1}, Value2: ${atomicObject.value2}, Value3: ${atomicObject.value3}`);
        }
        else{
            console.log("SUBSCRIBER : No value in queue");
        }

        /*--------- Example 2 : Wait for notification with timeout ---------*/
        atomicObject = subscriber.readWaitMS(2000);
        // Verify if the queue had a value on notification.
        if(atomicObject !== null && atomicObject !== undefined){
            console.log(`SUBSCRIBER : Value1: ${atomicObject.value1}, Value2: ${atomicObject.value2}, Value3: ${atomicObject.value3}`);
        }
        else{
            console.log("SUBSCRIBER : No value in queue");
        }
    }
}

loop().catch(console.error);